frontend/src/metabase/common/components/ActionButton/ActionButton.tsx

```diff
@@ -0,0 +1,152 @@
+};
+
+export const ActionButton = forwardRef<ActionButtonHandle, ActionButtonProps>(
+  function ActionButtonInner(
+    {
+      normalText = t`Save`,
+      activeText = t`Saving...`,
+      failedText = t`Save failed`,
+      successText = t`Saved`,
+      useLoadingSpinner = false,
+      actionFn,
+      className = "",
+      children,
+      innerRef,
+      ...buttonProps
+    },
+    ref,
+  ) {
+    const [active, setActive] = useState(false);
+    const [result, setResult] = useState<null | "failed" | "success">(null);
+    const timeout = useRef<null | ReturnType<typeof setTimeout>>(null);
+    const actionPromise = useRef<CancellablePromise<unknown> | null>(null);
+
+    const resetTimeout = () => {
+      timeout.current && clearTimeout(timeout.current);
+    };
+
+    useEffect(() => {
+      const promise = actionPromise.current;
+      return () => {
+        resetTimeout();
+        if (promise) {
+          promise?.cancel?.();
+        }
+      };
+    }, [actionPromise]);
+
+    const resetState = useCallback(() => {
+      resetTimeout();
+      setActive(false);
+      setResult(null);
+    }, []);
+
+    // very rarely, we need to reset a button based on changes outside this button
+    useImperativeHandle(ref, () => ({ resetState }), [resetState]);
+
+    const resetStateOnTimeout = () => {
+      // clear any previously set timeouts then start a new one
+      resetTimeout();
+      timeout.current = setTimeout(resetState, 5000);
+    };
+
+    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+      event.preventDefault();
+
+      // set state to active
+      setActive(true);
+      setResult(null);
+
+      // run the function we want bound to this button
+      actionPromise.current = cancelable(actionFn());
+      actionPromise.current.then(
+        () => {
+          setActive(false);
+          setResult("success");
+          resetStateOnTimeout();
+        },
+        (error: Error & { isCanceled?: boolean }) => {
+          if (!error.isCanceled) {
+            console.error(error);
+            setActive(false);
+            setResult("failed");
+            resetStateOnTimeout();
+          }
+        },
+      );
+    };
+    const isActionDisabled = active || result === "success";
+    const actionStatus = active ? "pending" : (result ?? "idle");
+
+    const color = match({ result })
+      .with({ result: "success" }, () => "success")
+      .with({ result: "failed" }, () => "danger")
+      .otherwise(() => buttonProps.color || "brand");
+
+    // success and failure are always "filled" variant
+    const variant = match({ result })
+      .with({ result: "success" }, () => "filled")
+      .with({ result: "failed" }, () => "filled")
+      .otherwise(() => buttonProps.variant || "filled");
+
+    return (
+      <Button
+        loading={active && useLoadingSpinner}
+        ref={innerRef}
+        data-action-status={actionStatus}
+        {...buttonProps}
```

Comment:
`butttonProps` has `ButtonProps` type, and `ButtonProps` has it's own `color` and `variant` props. 
But `color` and `variant` identifiers are already bounded to `buttonProps`. That means that the priority of applying these props isn't strait forward. I mean, we need to know that `{ ... buttonProps }` must go before `color` and `variant` props because they are overrided somewhere above.